src/tools/illink/src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -76,17 +77,79 @@ public override void Initialize (AnalysisContext context)
 					if (context.OwningSymbol.IsInRequiresUnreferencedCodeAttributeScope (out _))
 						return;
 
-
 					foreach (var operationBlock in context.OperationBlocks) {
 						TrimDataFlowAnalysis trimDataFlowAnalysis = new (context, operationBlock);
 						trimDataFlowAnalysis.InterproceduralAnalyze ();
 						foreach (var diagnostic in trimDataFlowAnalysis.TrimAnalysisPatterns.CollectDiagnostics ())
 							context.ReportDiagnostic (diagnostic);
 					}
 				});
-				context.RegisterSyntaxNodeAction (context => {
-					ProcessGenericParameters (context);
-				}, SyntaxKind.GenericName);
+				// Examine generic instantiations in base types and interface list
+				context.RegisterSymbolAction (context => {
+					var type = (INamedTypeSymbol) context.Symbol;
+					if (type.BaseType is INamedTypeSymbol baseType) {
+						foreach (var diagnostic in ProcessGenericParameters (baseType, type.Locations[0]))
+							context.ReportDiagnostic (diagnostic);
+					}
+
+					foreach (var interfaceType in type.Interfaces) {
+						foreach (var diagnostic in ProcessGenericParameters (interfaceType, type.Locations[0]))
+							context.ReportDiagnostic (diagnostic);
+					}
+				}, SymbolKind.NamedType);
+				// Examine generic instantiations in method return type and parameters.
+				// This includes property getters and setters.
+				context.RegisterSymbolAction (context => {
+					var method = (IMethodSymbol) context.Symbol;
+					if (method.IsInRequiresUnreferencedCodeAttributeScope (out _))
+						return;
+
+					var returnType = method.ReturnType;
+					foreach (var diagnostic in ProcessGenericParameters (returnType, method.Locations[0]))
+						context.ReportDiagnostic (diagnostic);
+
+					foreach (var parameter in method.Parameters) {
+						foreach (var diagnostic in ProcessGenericParameters (parameter.Type, parameter.Locations[0]))
+							context.ReportDiagnostic (diagnostic);
+					}
+				}, SymbolKind.Method);
+				// Examine generic instantiations in field type.
+				context.RegisterSymbolAction (context => {
+					var field = (IFieldSymbol) context.Symbol;
+					if (field.IsInRequiresUnreferencedCodeAttributeScope (out _))
+						return;
+
+					foreach (var diagnostic in ProcessGenericParameters (field.Type, field.Locations[0]))
+						context.ReportDiagnostic (diagnostic);
+				}, SymbolKind.Field);
+				// Examine generic instantiations in invocations, of generically instantiated methods,
+				// or methods on generically instantiated types.
+				context.RegisterOperationAction (context => {
+					if (context.ContainingSymbol.IsInRequiresUnreferencedCodeAttributeScope (out _))
+						return;
+
+					var invocation = (IInvocationOperation) context.Operation;
+					var methodSymbol = invocation.TargetMethod;
+					foreach (var diagnostic in ProcessGenericParameters (methodSymbol, invocation.Syntax.GetLocation()))
+						context.ReportDiagnostic (diagnostic);
```

Comment:
Just curious - is there a reason to use this pattern of returning a list of diagnostics and iterating over it just to call report. Why not call report inside the method? It seems there's lot of unnecessary code duplication this way.

Reply:
Now I remember hitting this myself - it's a shame the context doesn't implement an interface with the parts which are common across all contexts (like `ReportDiagnostics`).